### PR TITLE
[dist][sharded_tensor] Fix ChunkShardingSpec metadata offsets for empty shards

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121002

ChunkShardingSpec generated metadata where offsets exceed the tensor size.

Example:

Torchrec prepared ShardedTensorMetadata:
```
ShardedTensorMetadata(shards_metadata=[
ShardMetadata(shard_offsets=[0, 0], shard_sizes=[2, 512], placement=rank:0/cuda:0), 
ShardMetadata(shard_offsets=[2, 0], shard_sizes=[2, 512], placement=rank:1/cuda:1), 
ShardMetadata(shard_offsets=[4, 0], shard_sizes=[2, 512], placement=rank:2/cuda:2), 
ShardMetadata(shard_offsets=[6, 0], shard_sizes=[2, 512], placement=rank:3/cuda:3), 
ShardMetadata(shard_offsets=[8, 0], shard_sizes=[2, 512], placement=rank:4/cuda:4), 
ShardMetadata(shard_offsets=[10, 0], shard_sizes=[0, 512], placement=rank:5/cuda:5), 
ShardMetadata(shard_offsets=[10, 0], shard_sizes=[0, 512], placement=rank:6/cuda:6)
],
size=torch.Size([10, 512]
), 
```
Calling ShardedTensor._init_from_local_shards_and_global_metadata()
ShardedTensor ShardingSpec builds metadata

```
ShardedTensorMetadata(shards_metadata=[
ShardMetadata(shard_offsets=[0, 0], shard_sizes=[2, 512], placement=rank:0/cuda:0), 
ShardMetadata(shard_offsets=[2, 0], shard_sizes=[2, 512], placement=rank:1/cuda:1), 
ShardMetadata(shard_offsets=[4, 0], shard_sizes=[2, 512], placement=rank:2/cuda:2), 
ShardMetadata(shard_offsets=[6, 0], shard_sizes=[2, 512], placement=rank:3/cuda:3), 
ShardMetadata(shard_offsets=[8, 0], shard_sizes=[2, 512], placement=rank:4/cuda:4), 
ShardMetadata(shard_offsets=[10, 0], shard_sizes=[0, 512], placement=rank:5/cuda:5), 
ShardMetadata(shard_offsets=[12, 0], shard_sizes=[0, 512], placement=rank:6/cuda:6)
],
size=torch.Size([10, 512]), tensor_properties=TensorProperties(dtype=torch.float16, layout=torch.strided, requires_grad=False, memory_format=torch.contiguous_format, pin_memory=False))
```
The deduced ChunkShardingSpec:
```
ChunkShardingSpec(dim=0, placements=[rank:0/cuda:0, rank:1/cuda:1, rank:2/cuda:2, rank:3/cuda:3, rank:4/cuda:4, rank:5/cuda:5, rank:6/cuda:6])
```

The fix is to limit offsets by dim size.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Differential Revision: [D54419513](https://our.internmc.facebook.com/intern/diff/D54419513)

### CODE REVIEW

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def build_metadata(self,
             chunked_dim_size = get_chunked_dim_size(sharding_dim_size, split_size, idx)
             shard_size = list(tensor_sizes)
             current_offsets = [0] * tensor_num_dim
-            current_offsets[self.dim] = split_size * idx  # type: ignore[index]
+            current_offsets[self.dim] = min(sharding_dim_size, split_size * idx)  # type: ignore[index]
```
Comment (colin2328) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:96]: nice, thanks for finding @IvanKobzarev 

@wanchaol , look good?

Comment (yifuwang) [inline torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:96]: `ChunkShardingSpec` says:
```
... it represents sharding a Tensor along a single dimension into equal chunks (similar to :meth:`torch.chunk`).
```

Not sure changing this would break systems (e.g. Model Store?) that relies on the property.

Any chance `EnumerableShardingSpec` would work for this case?

Comment (IvanKobzarev) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:96]: `ChunkShardingSpec`  was deduced in `ShardedTensor.init_from_local_shards_and_global_metadata`.
If `ChunkShardingSpec` was not supposed to be used with uneven shards(but sizes are already calculated uneven), we need to change the deduction of Spec in `_infer_sharding_spec_from_shards_metadata` https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharding_spec/api.py#L169

Comment (yifuwang) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:96]: A that's useful info. Do you think adding a condition for checking uniform shard size here would work? https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharding_spec/api.py#L212

My memory is a bit rusty on this, does our infra support `EnumerableShardingSpec`?

Comment (IvanKobzarev) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:96]: I do not know about EnumerableShardingSpec support.
But empty and uneven shards are valid for Torchrec usecase and used.
I will add offsets comparison to spec deduction.

**Comment by IvanKobzarev** [comment]
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54419513).

**Review by wz337 — APPROVED** [review]
LGTM! Thanks for the fix.

**Review by wz337 — APPROVED** [review]

File: torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py (hunk 1 above)
Comment (wz337) [reply torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py]: @yifuwang I don't know too much historical context on the original design, but it seems a little bit restrictive to allow ChunkShardingSpec to be uniform. I just checked for FSDP this issue doesn't appear, because the metatdata (size, offsets ...) is manually being created here. https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/_shard_utils.py#L26

For the uneven-sharded case, it seems ShardedTensor is ok with taking the uneven offsets and sizes and generating ChunkShardingSpec behind the scenes. 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_shard/sharded_tensor/api.py#L721

Therefore, I think this fix should be fine, but we can wait for the sandcastle tests to confirm that it doesn't break anything on modelstore.
